proposals/NNNN-outofproc-compiler-api-architecture.md

```diff
@@ -0,0 +1,408 @@
+immediately calls the c api provided include handler callback. The results are
+collected and new "continue compilation" `Request` is dispatched.
+
+The worker process (currently waiting for another `Request` to continue)
+receives the `Request` with include handler data as parameters and proceeds
+with compilation.
+
+This `Request`/`Response` pattern continues until the compilation is completed.
+Compilation is always concluded with a final `Response` closing the
+`Request`/`Response` pair.
+
+```
+REQUEST  - "compile" (include handler support is enabled)
+
+( compiler needs an include, calls IPC wrapped included handler callback )
+
+RESPONSE - "needinclude"
+
+( originating caller's callback gets called and results are collected and
+passed into a new request to continue compilation )
+
+REQUEST - "continuecompile" (include handler result passed as params)
+RESPONSE - "needinclude"
+
+( originating caller's callback gets called and results are collected and
+passed into a new request to continue compilation )
+
+REQUEST  - "continuecompile"
+
+( no more includes are encountered and compilation reaches the end )
+
+RESPONSE - "result", compilation succeeded/failed.
+```
+
+The following is a JSON-RPC example of this pattern.
+
+### Example: compile, include handler, success
+```
+**** start compilation ****
+--> {"jsonrpc": "2.0", "method": "compile", "params": {"arg1": 23, "arg2": 42},
+     "includehandler", "true", "id": 1}
+
+**** include needed, so response indicates this and supplies a path ****
+<-- {"jsonrpc": "2.0", "result": {"needinclude": "true", "path": "somepath"}, "id": 1}
+
+**** continue compilation, supplying include data ****
+--> {"jsonrpc": "2.0", "method": "continuecompile", "params": {"path": "somepath"}, "id": 2}
+
+**** include needed, so response indicates this and supplies a path ****
+<-- {"jsonrpc": "2.0", "result": {"needinclude": "true", "path": "somepath2"}, "id": 2}
+
+**** continue compilation, supplying include data ****
+--> {"jsonrpc": "2.0", "method": "continuecompile", "params": {"path": "somepath"}, "id": 3}
+
+**** compilation completed ***
+<-- {"jsonrpc": "2.0", "result": {"stdout": "somepath", "stderr": "somepath"},
+     "id": 3}
+```
+
+### The worker process
+The compiler driver code for DXC lives in clang.exe.  This module will be
+extended with additional command-line arguments and launch behaviors.
+Additional params will be used to configure the process startup logic to setup
+the required IPC for communicating back to the thread that launched it.
+
+Using this same module keeps a single shipping binary for all compilation.
+
+## Alternatives considered
+An alternate design could be to reduce the total monitor threads down by
+using one thread to monitor all clang.exe processes. This would centralize all
+rpc traffic into a single place, but could bottleneck performance if 
+operations are waiting on workers to respond.
```

Comment:
Why was this alternative rejected?

> could bottleneck performance if operations are waiting on workers to respond

This would only happen if we had synchronous communication with the workers.

Reply:
Yeah.. I changed my mind on this.  I wasn't really thinking through out the dispatcher could be built so a monitor thread for multiple workers is totally doable.